From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Matteo Bruni <mbruni@codeweavers.com>
Date: Tue, 27 Aug 2019 02:55:16 +0200
Subject: [PATCH] wined3d: Avoid NaNs in LOG and LOGP.

CW-Bug-Id: #21762

Fixes some white artifacts in A Hat in Time Chapter 2 Act 1.
Valve-commit: a054ab5b62b9120e80056a9beaca7ff71a340d80
Picked: clean
Class: valve
Evidence: docs/evidence/2026-09-28-wine-proton-rebase.md
Offered-upstream: no
---
 dlls/wined3d/glsl_shader.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/dlls/wined3d/glsl_shader.c b/dlls/wined3d/glsl_shader.c
index d7ae993..02e4f5c 100644
--- a/dlls/wined3d/glsl_shader.c
+++ b/dlls/wined3d/glsl_shader.c
@@ -4443,7 +4443,10 @@ static void shader_glsl_scalar_op(const struct wined3d_shader_instruction *ins)
 
         case WINED3DSIH_LOG:
         case WINED3DSIH_LOGP:
-            string_buffer_sprintf(prefix, "log2(abs(");
+            if (shader_version <= WINED3D_SHADER_VERSION(3, 0))
+                string_buffer_sprintf(prefix, "%s == 0.0 ? -FLT_MAX : log2(abs(", src0_param.param_str);
+            else
+                string_buffer_sprintf(prefix, "log2(abs(");
             string_buffer_sprintf(suffix, "))");
             break;
 
